projects/chaos/src/sequencer.c

```diff
@@ -0,0 +1,365 @@
+    .next = { .id = CHAOS_EVENT_OPEN_RELAY, .data = RELAY_ID_BATTERY },
+    .await = true },
+  { .expected = { .id = CHAOS_EVENT_OPEN_RELAY, .data = RELAY_ID_BATTERY },
+    .next = SEQUENCER_NO_OP,
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_RELAY_OPENED, .data = RELAY_ID_BATTERY },
+    .next = { .id = CHAOS_EVENT_GPIO_IDLE, .data = SEQUENCER_EMPTY_DATA },
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_GPIO_IDLE, .data = SEQUENCER_EMPTY_DATA },
+    .next = { .id = CHAOS_EVENT_SEQUENCE_IDLE_DONE, .data = SEQUENCER_EMPTY_DATA },
+    .await = false },
+};
+
+// Charge State Order:
+// - Turn on charging GPIOs
+// - Open battery relay
+// - Open solar master rear relay
+// - Open solar master front relay
+// - Open main power relay
+// - Enable DCDC monitoring
+static const SequencerEventPair s_charge_events[] = {
+  { .expected = SEQUENCER_NO_OP,
+    .next = { .id = CHAOS_EVENT_GPIO_CHARGE, .data = SEQUENCER_EMPTY_DATA },
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_GPIO_CHARGE, .data = SEQUENCER_EMPTY_DATA },
+    .next = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_BATTERY },
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_BATTERY },
+    .next = SEQUENCER_NO_OP,
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_BATTERY },
+    .next = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_REAR },
+    .await = true },
+  { .expected = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_REAR },
+    .next = SEQUENCER_NO_OP,
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_SOLAR_MASTER_REAR },
+    .next = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_FRONT },
+    .await = true },
+  { .expected = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_FRONT },
+    .next = SEQUENCER_NO_OP,
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_SOLAR_MASTER_FRONT },
+    .next = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_MAIN_POWER },
+    .await = true },
+  { .expected = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_MAIN_POWER },
+    .next = SEQUENCER_NO_OP,
+    .await = false },
+  { .expected = { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_MAIN_POWER },
+    .next = { .id = CHAOS_EVENT_MONITOR_ENABLE, .data = POWER_PATH_SOURCE_ID_DCDC },
+    .await = true },
+  { .expected = { .id = CHAOS_EVENT_MONITOR_ENABLE, .data = POWER_PATH_SOURCE_ID_DCDC },
+    .next = { .id = CHAOS_EVENT_SEQUENCE_CHARGE_DONE, .data = SEQUENCER_EMPTY_DATA },
+    .await = false },
+};
+
+// Drive State Order:
+// - Turn on drive GPIOs
+// - Open battery relay
+// - Open solar master rear relay
+// - Open solar master front relay
+// - Open main power relay
+// - Enable DCDC monitoring
+static const SequencerEventPair s_drive_events[] = {
+  { .expected = SEQUENCER_NO_OP,
```

Comment:
It seems like you're duplicating events here. Could you use "raise/response" pairs instead? I think the current model dictates something more along the lines of "previous/next" pairs.

i.e. something like
```
{ .raise = { .id = CHAOS_EVENT_GPIO_DRIVE, .data = ... } },
{ .raise = { .id = CHAOS_EVENT_CLOSE_RELAY, .data = ... }, .response = { .id = CHAOS_EVENT_RELAY_CLOSED, .data = ... } }
```
I believe you could look at the previous pair to see if the current event is "expected" to match the current behavior.

Reply:
Done. Good call this is much nicer.